src/Compilers/CSharp/Test/Symbol/Symbols/ModuleInitializersTests.cs

```diff
@@ -0,0 +1,108 @@
+
+        [Fact]
+        public static void ModuleInitializersNotUsableInCSharp8()
+        {
+            var source =
+@"using System.Runtime.CompilerServices;
+
+class C
+{
+    [ModuleInitializer]
+    internal static void M() { }
+}
+
+namespace System.Runtime.CompilerServices { class ModuleInitializerAttribute : System.Attribute { } }
+";
+            var compilation = CreateCompilation(source, parseOptions: TestOptions.Regular8);
+
+            compilation.VerifyDiagnostics(
+                // (5,6): error CS8652: The feature 'module initializers' is currently in Preview and *unsupported*. To use Preview features, use the 'preview' language version.
+                //     [ModuleInitializer]
+                Diagnostic(ErrorCode.ERR_FeatureInPreview, "ModuleInitializer").WithArguments("module initializers").WithLocation(5, 6)
+                );
+        }
+
+        [Fact]
+        public void ModuleTypeStaticConstructorIsNotEmittedWhenNoMethodIsMarkedWithModuleInitializerAttribute()
+        {
+            string source = @"
+using System;
+using System.Runtime.CompilerServices;
+
+class C
+{
+    internal static void M() => Console.WriteLine(""C.M"");
+}
+
+class Program 
+{
+    static void Main() => Console.WriteLine(""Program.Main"");
+}
+
+namespace System.Runtime.CompilerServices { class ModuleInitializerAttribute : System.Attribute { } }
+";
+
+            CompileAndVerify(
+                source,
+                parseOptions: s_parseOptions,
+                options: new CSharpCompilationOptions(OutputKind.ConsoleApplication, metadataImportOptions: MetadataImportOptions.All),
+                symbolValidator: module =>
+                {
+                    var rootModuleType = (TypeSymbol)module.GlobalNamespace.GetMember("<Module>");
+                    Assert.Null(rootModuleType.GetMember(".cctor"));
+                },
+                expectedOutput: @"
+Program.Main");
+        }
+
+        [Fact]
+        public void ModuleTypeStaticConstructorCallsMethodMarkedWithModuleInitializerAttribute()
+        {
+            string source = @"
+using System;
+using System.Runtime.CompilerServices;
+
+class C
+{
+    [ModuleInitializer]
+    internal static void M() => Console.WriteLine(""C.M"");
+}
+
+class Program 
+{
+    static void Main() => Console.WriteLine(""Program.Main"");
+}
+
+namespace System.Runtime.CompilerServices { class ModuleInitializerAttribute : System.Attribute { } }
+";
+
+            var x = CompileAndVerify(
+                source,
+                parseOptions: s_parseOptions,
+                options: new CSharpCompilationOptions(OutputKind.ConsoleApplication, metadataImportOptions: MetadataImportOptions.All),
+                symbolValidator: module =>
+                {
+                    var rootModuleType = (TypeSymbol)module.GlobalNamespace.GetMember("<Module>");
+                    Assert.NotNull(rootModuleType.GetMember(".cctor"));
```

Comment:
>Assert.NotNull(rootModuleType.GetMember(".cctor")); [](start = 20, length = 51)

Please assert other properties of this symbol that are specified by ```RootModuleStaticConstructor```. Like, special name, not new slot, etc.

Reply:
Since the symbols being validated have been roundtripped through IL, can I use `Assert.Equal(MethodKind.StaticConstructor, staticConstructor.MethodKind);` in place of things like new slot/abstract/sealed etc?

Reply:
> Since the symbols being validated have been roundtripped through IL, can I use Assert.Equal(MethodKind.StaticConstructor, staticConstructor.MethodKind); in place of things like new slot/abstract/sealed etc?

I would prefer us to go after specific properties simply because it is unlikely we validate much when we calculate method kind. Asserting it, is a good idea as well. We are dealing with PEMethodSymbol, it is likely to have flags as they come from metadata, it is fine to assert that if you find that more convenient.


---
In reply to: [414888423](https://github.com/dotnet/roslyn/pull/43281#discussion_r414888423) [](ancestors = 414888423)